<?xml version="1.0"?>
<robot name="robot"> 

  <link name="world"/>
       
<link name="base_link">
    <inertial>
        <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
        <mass value="4.0"/>
        <inertia ixx="0.223" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.223" izz="0.02"/>
    </inertial>

    <visual>
        <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
        <geometry>
            <cylinder radius="0.1" length="0.8"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
         <origin xyz="0.0 0.0 0.4" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="0.1" length="0.8"/>
        </geometry>
    </collision>
</link>


<joint name="fixed1" type="fixed">
    <parent link="world"/>
    <child link="link_S1"/>
</joint>

<link name="link_1">
    <inertial>
        <origin xyz="0.3 0 0" rpy="0 1.5705 0.0" />
        <mass value="2.0"/>
        <inertia ixx="0.1116" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.1116" izz="0.01"/>
    </inertial>

    <visual>
        <origin xyz="0.3 0 0" rpy="0 1.5705 0.0" />
        <geometry>
            <cylinder radius="0.1" length="0.8"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
     <origin xyz="0.3 0 0" rpy="0 1.5705 0.0" />
        <geometry>
            <cylinder radius="0.1" length="0.8"/>
        </geometry>
    </collision>
</link>

<link name="link_S1">
    <inertial>
        <origin xyz="-1.0 0.0 0.575" rpy="0 0 0.0" />
        <mass value="1.0"/>
        <inertia ixx="0.1116" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.1116" izz="0.01"/>
    </inertial>

    <visual>
        <origin xyz="-1.0 0.0 0.575" rpy="0 0 0.0" />
        <geometry>
            <cylinder radius="0.15" length="0.15"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
       <origin xyz="-1.0 0.0 0.575" rpy="0 0 0.0" />
        <geometry>
            <cylinder radius="0.15" length="0.15"/>
        </geometry>
    </collision>
</link>

<link name="link_S2">
    <inertial>
      <origin xyz="-1 0 0.25" rpy="0 0 0.0"/>
        <mass value="5"/>
       <inertia ixx="0.44166" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.44166" izz="0.208333"/>
    </inertial>

    <visual>
        <origin xyz="-1 0 0.25" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.5 0.9 0.5"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
    <origin xyz="-1 0 0.25" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.03 0.12 0.05"/>
        </geometry>
    </collision>
</link>
<joint name="fixedS" type="fixed">
    <parent link="link_S1"/>
    <child link="link_S2"/>
</joint>

<joint name="joint_1" type="revolute">
    <parent link="link_S1"/>
    <child link="base_link"/>
    <axis xyz="0 0 1"/>
    <limit lower="-6.28318" upper="6.28318" velocity="50.0" effort="1000.0"/>
    <origin xyz="-1 0 0.5" rpy="0 0 0"/>
</joint>
<joint name="joint_2" type="prismatic">
    <parent link="base_link"/>
    <child link="link_1"/>
    <axis xyz="1 0 0"/>
    <limit lower="-0.5" upper="0" velocity="0.5" effort="1000.0"/>
    <origin xyz="0.0 0 0.7" rpy="0 0 0"/>
</joint>
<link name="link_2">
    <inertial>
         <origin xyz="-0.3 0 0.0" rpy="0 1.5705 0.0" />
        <mass value="1.0"/>
        <inertia ixx="0.0547" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.0547" izz="0.0028125"/>
    </inertial>

    <visual>
        <origin xyz="-0.3 0 0.0" rpy="0 1.5705 0.0" />
        <geometry>
            <cylinder radius="0.075" length="0.8"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
          <origin xyz="-0.3 0 0.0" rpy="0 1.5705 0.0" />
        <geometry>
            <cylinder radius="0.075" length="0.8"/>
        </geometry>
    </collision>
</link>
 
<joint name="joint_3" type="prismatic">
    <parent link="link_1"/>
    <child link="link_2"/>
    <axis xyz="1 0 0"/>
    <limit lower="0" upper="0.5" velocity="0.5" effort="1000.0"/>
    <origin xyz="0.8 0 0" rpy="0 0 0"/>
</joint>

 
<link name="link_3">
    <inertial>
      <origin xyz="0.0 0 -0.075" rpy="0 0 0" />
        <mass value="0.5"/>
        <inertia ixx="0.004453" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.004453" izz="0.00140625"/>   
        </inertial>

    <visual>
        <origin xyz="0.0 0 -0.075" rpy="0 0 0" />
        <geometry>
            <cylinder radius="0.075" length="0.3"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
     <origin xyz="0.0 0 -0.075" rpy="0 0 0" />
        <geometry>
            <cylinder radius="0.075" length="0.3"/>
        </geometry>
    </collision>
</link>

<joint name="joint_4" type="revolute">
    <parent link="link_2"/>
    <child link="link_3"/>
    <axis xyz="0 0 1"/>
    <limit lower="-3.14159" upper="3.14159" velocity="50.0" effort="1000.0"/>
    <origin xyz="0 0 0.0" rpy="0 0 0"/>
</joint>

<link name="link_4">
    <inertial>
        <origin xyz="0.0 0 -0.025" rpy="0 0 1.5707" />
        <mass value="0.5"/>
        <inertia ixx="0.006979" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.006979" izz="0.00625"/>
    </inertial>

    <visual>
        <origin xyz="0.0 0 -0.025" rpy="0 0 1.5707" />
        <geometry>
            <cylinder radius="0.05" length="0.4"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
      <origin xyz="0.0 0 -0.025" rpy="0 0 1.5707" />
        <geometry>
            <cylinder radius="0.05" length="0.4"/>
        </geometry>
    </collision>
</link>
 
<joint name="joint_5" type="prismatic">
    <parent link="link_3"/>
    <child link="link_4"/>
    <axis xyz="0 0 -1"/>
    <limit lower="0.0" upper="0.395" velocity="0.5" effort="1000.0"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <dynamics damping="2.0" friction="0.0"/>
    <dynamics damping="2.0" friction="0.0"/>
</joint>

<link name="link_5">
    <inertial>
           <origin xyz="0.8 0.0 0.4625" rpy="0 0 0.0"/>
        <mass value="0.5"/>
        <inertia ixx="0.0006375" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.003204" izz="0.00264166"/>
    </inertial>


    <visual>
        <origin xyz="0.8 0.0 0.4625" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.25 0.12 0.03"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
         <origin xyz="0.8 0.0 0.4625" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.25 0.12 0.03"/>
        </geometry>
    </collision>
</link>
<joint name="fixed2" type="fixed">
    <parent link="link_4"/>
    <child link="link_5"/>
    <origin xyz="-0.8 0.0 -0.7" rpy="0.0 0.0 0.0"/>
</joint>
<link name="link_6">
    <inertial>
        <origin xyz="0.69 0.0 0.4325" rpy="0 0 0.0"/>
        <mass value="0.25"/>
        <inertia ixx="0.000375" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.00031875" izz="0.00009375"/>
    </inertial>


    <visual>
        <origin xyz="0.69 0.0 0.4325" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.03 0.12 0.05"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
      <origin xyz="0.69 0.0 0.4325" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.03 0.12 0.05"/>
        </geometry>
    </collision>
</link>
<joint name="joint_6" type="prismatic">
    <parent link="link_5"/>
    <child link="link_6"/>
    <axis xyz="1 0 0"/>
    <limit lower="0" upper="0.09" velocity="0.5" effort="1000.0"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
     <dynamics damping="2.0" friction="0.0"/>
     <dynamics damping="2.0" friction="0.0"/>

</joint>

<link name="link_7">
    <inertial>
      <origin xyz="0.91 0.0 0.4325" rpy="0 0 0.0"/>
        <mass value="0.25"/>
       <inertia ixx="0.000375" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.00031875" izz="0.00009375"/>
    </inertial>

    <visual>
        <origin xyz="0.91 0.0 0.4325" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.03 0.12 0.05"/>
        </geometry>

        <material name="silver">
            <color rgba="0.0 0.7 0.0 0.7"/>
        </material>
    </visual>
    
    <collision>
   <origin xyz="0.91 0.0 0.4325" rpy="0 0 0.0"/>
        <geometry>
            <box size="0.03 0.12 0.05"/>
        </geometry>
    </collision>
</link>
<joint name="joint_7" type="prismatic">
    <parent link="link_5"/>
    <child link="link_7"/>
    <axis xyz="-1 0 0"/>
    <limit lower="0" upper="0.09" velocity="0.5" effort="1000.0" />
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <dynamics damping="0.5" friction="0.0"/>
       
      
</joint>

  <joint name="camera_joint" type="fixed">
        <parent link="world"/>
        <child link="camera_link"/>
        <origin xyz="0 0 2.2" rpy="0 1.57 0"/>
    </joint>

    <link name="camera_link">
        <inertial>
         <origin xyz="0 0 0.0"/>
         <mass value="0.25"/>
         <inertia ixx="0.000375" ixy="0.0" ixz="0.0" iyz="0.0"  iyy="0.00031875" izz="0.00009375"/>
        </inertial>

        <visual>
         <origin xyz="0 0 0.0"/> 
            <geometry>
                <box size="0.05 0.05 0.05" />
            </geometry>
            <material name="silver">
            <color rgba="1.0 0 0.0 0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0.0"/>
            <geometry>
                 <box size="0.05 0.05 0.05"/>
            </geometry>
            <material name="silver">
            <color rgba="1.0 0 0.0 0"/>
            </material>
        </collision>
    </link>


    <joint name="camera_optical_joint" type="fixed">
        <parent link="camera_link"/>
        <child link="camera_link_optical"/>
        <origin xyz="0 0 0" rpy="1.57 0 1.57"/>
    </joint>

    <link name="camera_link_optical"></link>



    <gazebo reference="camera_link">
        <material>Gazebo/silver</material>

        <sensor name="camera" type="camera">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>true</visualize>
            <update_rate>5</update_rate>
            <camera>
                <horizontal_fov>1.089</horizontal_fov>
                <image>
                    <format>R8G8B8</format>
                    <width>420</width>
                    <height>420</height>
                </image>
                <clip>
                    <near>0.05</near>
                    <far>1.5</far>
                </clip>
                     <noise>
                      <type>gaussian</type>
                        <!-- Noise is sampled independently per pixel on each frame.
                          That pixel's noise value is added to each of its color
                          channels, which at that point lie in the range [0,1]. -->
                     <mean>0.0</mean>
                     <stddev>0.007</stddev>
                    </noise>
            </camera>
            <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
                 <alwaysOn>true</alwaysOn>
                 <updateRate>0.0</updateRate>
                 <cameraName>robot/camera1</cameraName>
                 <imageTopicName>image_raw</imageTopicName>
                 <cameraInfoTopicName>camera_info</cameraInfoTopicName>
                 <frame_name>camera_link_optical</frame_name>
                 <hackBaseline>0.07</hackBaseline>
                 <distortionK1>0.0</distortionK1>
                 <distortionK2>0.0</distortionK2>
                 <distortionK3>0.0</distortionK3>
                   <distortionT1>0.0</distortionT1>
                 <distortionT2>0.0</distortionT2>
            </plugin>
        </sensor>
    </gazebo>




<transmission name ="trans_1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_1">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_1">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>

<transmission name ="trans_2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_2">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_2">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>

<transmission name ="trans_3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_3">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_3">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>
<transmission name ="trans_4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_4">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_4">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>

<transmission name ="trans_5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_5">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_5">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>

<transmission name ="trans_6">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_6">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_6">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>

<transmission name ="trans_7">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_7">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_7">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
    </actuator>
</transmission>
<gazebo>
    <plugin name="gazebo_grasp_fix" filename="libgazebo_grasp_fix.so">
       <arm>
         <arm_name>robot</arm_name>
         <palm_link>link_4</palm_link>
         <gripper_link>link_6</gripper_link>
         <gripper_link>link_7</gripper_link>
        </arm>
       <forces_angle_tolerance>100</forces_angle_tolerance>
       <update_rate>10</update_rate>
       <grip_count_threshold>3</grip_count_threshold>
       <max_grip_count>10</max_grip_count>
       <release_tolerance>0.001</release_tolerance>
       <disable_collisions_on_attach>false</disable_collisions_on_attach>
       <contact_topic>__default_topic__</contact_topic>
    </plugin>
</gazebo>

<gazebo>
    <plugin name="control" filename="libgazebo_ros_control.so"/>
</gazebo>




</robot>
